src/patterns.jl

```diff
@@ -0,0 +1,63 @@
+
+"""
+    Pattern
+
+A pattern is just a wrapper around a `Tuple{Vararg{Symbol}}` which enables searching and
+filtering for matching components and dimension paths in a [`Dataset`](@ref).
+Special symbols `:_` and `:__` are used as wildcards, similar to `*` and `**` in glob
+pattern matching.
+"""
+struct Pattern
+    segments::Tuple{Vararg{Symbol}}
+end
+
+Pattern(segments::Symbol...) = Pattern(segments)
+Base.convert(::Type{Pattern}, x::Tuple{Vararg{Symbol}}) = Pattern(x)
+
+# Primary functionality is to support identify tuples that match the constraint
+# NOTE: I'm not entirely sure if we should overload `in` for this algorithm
+function Base.in(item::Tuple{Vararg{Symbol}}, pattern::Pattern)
+    pat = pattern.segments
+    i = 1   # item  index
+    j = 1   # contraint index
+    n = length(item)
+    m = length(pat)
+
+    # If our constrain has more segments than the item then it isn't going to match
```

Comment:
```suggestion
    # If our constraint has more segments than the item then it isn't going to match
```

Reply:
Thanks, this comment was removed since I rewrote the `in` function to use `iterate`.